pkg/server/cstorvolumeconfig/restore_endpoint.go

```diff
@@ -0,0 +1,491 @@
+func (rOps *restoreAPIOps) getNodeID(scObj *storagev1.StorageClass) (string, error) {
+	var labelKey, labelValue string
+	cspcName := scObj.Parameters["cstorPoolCluster"]
+	if len(scObj.AllowedTopologies) != 0 {
+		labelKey = scObj.AllowedTopologies[0].MatchLabelExpressions[0].Key
+		labelValue = scObj.AllowedTopologies[0].MatchLabelExpressions[0].Values[0]
+	} else {
+		cspiList, err := rOps.clientset.CstorV1().
+			CStorPoolInstances(getOpenEBSNamespace()).
+			List(metav1.ListOptions{
+				LabelSelector: cstortypes.CStorPoolClusterLabelKey + "=" + cspcName,
+			})
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to list CSPI of CSPC %s", cspcName)
+		}
+		if len(cspiList.Items) == 0 {
+			return "", errors.Errorf("no cspi exists for CSPC %s", cspcName)
+		}
+		labelKey = cstortypes.HostNameLabelKey
+		labelValue = cspiList.Items[0].Spec.HostName
+	}
+
+	nodeList, err := rOps.k8sclientset.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: labelKey + "=" + labelValue})
+	if err != nil {
+		return "", errors.Wrapf(err, "falied to list nodes which has %s:%s label", labelKey, labelValue)
+	}
+	if len(nodeList.Items) == 0 {
+		return "", errors.Errorf("no nodes exists for provided label %s:%s", labelKey, labelValue)
+	}
+	return nodeList.Items[0].Name, nil
+}
+
+// buildCStorVolumeConfig build the CVC from StorageClass and restoreObject
+// NOTE: This function will be called only in case of local restore request
+func (rOps *restoreAPIOps) buildCStorVolumeConfig(
+	scObj *storagev1.StorageClass, restoreObj *openebsapis.CStorRestore) (*cstor.CStorVolumeConfig, error) {
+	nodeID, err := rOps.getNodeID(scObj)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get nodeID")
+	}
+	replicaCount, _ := strconv.Atoi(scObj.Parameters["replicaCount"])
+
+	// Build CStorVolumeConfig
+	// TODO: Convert all literals into constants and make as builder code
+	cvcObj := &cstor.CStorVolumeConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      restoreObj.Spec.VolumeName,
+			Namespace: getOpenEBSNamespace(),
+			Labels: map[string]string{
+				cstortypes.CStorPoolClusterLabelKey: scObj.Parameters["cstorPoolCluster"],
+				"openebs.io/source-volume":          restoreObj.Spec.VolumeName,
+			},
+			Annotations: map[string]string{
+				"openebs.io/volumeID":      restoreObj.Spec.VolumeName,
+				"openebs.io/volume-policy": scObj.Parameters["cstorVolumePolicy"],
+			},
+			Finalizers: []string{"cvc.openebs.io/finalizer"},
+		},
+		Spec: cstor.CStorVolumeConfigSpec{
+			Provision: cstor.VolumeProvision{
+				Capacity: corev1.ResourceList{
+					corev1.ResourceName(corev1.ResourceStorage): restoreObj.Spec.Size,
+				},
+				ReplicaCount: replicaCount,
+			},
+			Capacity: corev1.ResourceList{
+				corev1.ResourceName(corev1.ResourceStorage): restoreObj.Spec.Size,
+			},
+			// If CStorVolumeSource is mentioned then CVC controller treat it as a clone request
+			CStorVolumeSource: restoreObj.Spec.RestoreSrc + "@" + restoreObj.Spec.RestoreName,
```

Comment:
This should be reverse volumename@snapname i.e restoreObj.Spec.RestoreName + "@" + restoreObj.Spec.RestoreSrc

Reply:
Exiting one is correct.
```go
Spec.RestoreName: will hold snapshot name.
Spec.RestoreSrc: will hold source volume name.
```
```sh
curl -XPOST -d '{"metadata":{"namespace":"openebs","creationTimestamp":null},"spec":{"restoreName":"backup_snapshot","volumeName":"pvc-285eb80c-f23e-42ea-8136-8863c1c9eb47","restoreSrc":"pvc-185eb80c-f23e-42ea-8136-8863c1c9eb0e","storageClass":"openebs-csi-sc","size":"5Gi","localRestore":true},"status":""}'  http://10.101.149.30:5757/latest/restore/
{"iqn":"iqn.2016-09.com.openebs.cstor:pvc-285eb80c-f23e-42ea-8136-8863c1c9eb47","lun":0,"targetPortal":"10.100.35.120:3260"}
```

Reply:
no need of changing it.